### DISCUSSION TITLE
How to handle Python package dependencies of Python bindings

### DISCUSSION THREAD

**Discussion by aWZHY0yQH81uOYvH** [Homebrew/discussions#5001] [Writing Formulae/Casks]
I am working on a formula that primarily installs a C++ project from source. This C++ project has Python bindings, which I am also optionally installing with the formula. These bindings rely on other Python packages that must also be installed. Currently, I am following [this guide](https://docs.brew.sh/Python-for-Formula-Authors), which seems to indicate I should be putting specific resource stanzas in the formula and install them as follows.

```ruby
resource "h5py" do
  url "https://files.pythonhosted.org/packages/37/fc/0b1825077a1c4c79a13984c59997e4b36702962df0bca420698f77b70b10/h5py-3.10.0.tar.gz"
  sha256 "d93adc48ceeb33347eb24a634fb787efc7ae4644e6ea4ba733d099605045c049"
end

def install
  # Install dependencies into a virtual environment
  venv = virtualenv_create(libexec, "python3")
  venv.pip_install resources

  # Get python 3 sub-version we are currently using (3.x)
  python_version = Formula["python@3"].version.to_s.split(".")[0..1].join(".")

  # Create .pth file to reference packages in venv
  (lib/"python#{python_version}/site-packages/homebrew-xyz-dependencies.pth").write <<~EOS
    #{libexec}/lib/python#{python_version}/site-packages
  EOS
end
```

Although this works, it seems like a very ugly way of doing things, with hard-coded package versions and dependencies for those packages (including other Python and non-python dependencies). The compile time of all the package dependencies makes a `brew install` that previously took a little more than a minute now take nearly an hour.

This seems analogous to having a C++ project which depends on a giant library, and instead of simply installing the pre-compiled bottle, compiling the whole library from scratch and reserving its use for only one formula. However, the guide linked above seems to discourage using a non-venv `pip` for this purpose when dealing with Python packages (presumably since brew can't track what it does and generate an installation receipt).

With that said, is what I have above the actual recommended way of doing this? Is what I'm trying to do in the first place dumb?

**Comment by SMillerDev** [comment]
Why not depend on the numpy formula?

**Reply by aWZHY0yQH81uOYvH** [reply]
This is just a simplified (and evidently bad) example. I also need `h5py` and possibly `matplotlib`, which don't have Homebrew formulae.

(edited to refer to `h5py` instead of `numpy`)

**Comment by aWZHY0yQH81uOYvH** [comment]
Decided to go with [this](https://github.com/thliebig/openEMS-Project/pull/169/files#diff-f2505009946ca38d19250713ed5833e090bc713aa3a8044b867d7bf8cbaada45). There is now a `matplotlib` bottle, and I'm just installing `h5py` with `pip`. Since this formula isn't bottled, it has to be compiled with `--HEAD`, which seems to make `pip` still compile `h5py` from source, which takes much longer than compiling the actual project, but that at least seems more maintainable than using a fixed version that needs a new PR to update every time.
